global_buffer/global_buffer_bs_generator.py

```diff
@@ -0,0 +1,214 @@
+                    result.append(reg)
+        return result
+
+    def get_bitstream(self):
+        result = []
+        for core in self.cores:
+            for feature in core.features:
+                for reg in feature.registers.values():
+                    # If register value is default value, do not need to config
+                    if reg['default'] == reg['value']:
+                        continue
+                    result.append((reg['addr'], reg['value']))
+        return result
+
+class GlbCore(ABC):
+    def __init__(self, glb: "Glb", core_id: int):
+        self.glb = glb
+        self.core_id = core_id
+        self.features: List["GlbFeature"] = []
+        self.num_features = 0
+        self.feature_addr_width = self.glb.config_map.feature_addr_width
+        self.reg_addr_width = self.glb.config_map.reg_addr_width
+
+    def add_feature(self, _feature: "GlbFeature"):
+        if self.num_features >= 2**self.feature_addr_width:
+            raise ValueError(f"Cannot add more than {self.num_features}"
+                             f"features to the core")
+        self.features.append(_feature)
+        self.num_features = self.num_features + 1
+
+    @abstractmethod
+    def name(self):
+        pass
+
+
+class GlbFeature(ABC):
+    def __init__(self, core: "GlbCore", feature_id: int):
+        self.glb = core.glb
+        self.core = core
+        self.feature_id = feature_id
+        self.reg_addr_width = self.glb.config_map.reg_addr_width
+        self.feature_addr_width = self.glb.config_map.feature_addr_width
+        self.core_addr_width = self.glb.config_map.core_addr_width
+        self.config_data_width = self.glb.config_map.config_data_width
+        self.byte_addressable = self.glb.config_map.byte_addressable
+        if self.byte_addressable:
+            self.byte_offset =\
+                    int(math.ceil(math.log2(self.config_data_width/8)))
+        else:
+            self.byte_offset = 0
+        self.registers = DotDict()
+        self.num_registers = 0
+
+    def add_config(self, name, width, default_value=0):
+        if name in self.registers:
+            raise ValueError(f"{name} is already a register")
+        if self.num_registers >= 2**self.reg_addr_width:
+            raise ValueError(f"Cannot add more than {self.num_registers}"
+                             f"configuration registers to the feature")
+
+        reg_name = f"{self.name()}_{name}"
+        addr = (self.core.core_id << (self.feature_addr_width +
+                                      self.reg_addr_width +
+                                      self.byte_offset)) | \
+               (self.feature_id << (self.reg_addr_width +
+                                    self.byte_offset)) | \
+               (self.num_registers << self.byte_offset)
+        self.registers[f"{name}"] = {"name": reg_name, "addr": addr,
+                                     "range": width, "default": default_value,
```

Comment:
I usually interpret `range` as in `range()` in Python. I think you're using the actual width here. To be consistent with other configuration in the system, I'd suggest to use `range` here. If so you might also need to change the logic at line 204.

Reply:
To clarify, you'd suggest to use `width` or `range`?

Reply:
Use `range`. Interconnect doesn't have the concept of `width` and that's the reason why we should use `range` instead.

Reply:
got it.

Reply:
I think I misunderstood what you said.
So in interconnect, if the `range` of configuration register `A` is 2, does it mean that we can write `0`, `1`, and `2` to that register?
I don't see how it can be implemented in verilog.